test_conformance/printf/test_printf.cpp

```diff
@@ -214,15 +223,148 @@ int waitForEvent(cl_event* event)
 // Static helper functions definition
 //-----------------------------------------
 
+float get_random_float(float low, float high, MTdata d)
+{
+    float t = (float)((double)genrand_int32(d) / (double)0xFFFFFFFF);
+    return (1.0f - t) * low + t * high;
+}
+
+//-----------------------------------------
+// makeMixedFormatPrintfProgram
+//-----------------------------------------
+cl_program makeMixedFormatPrintfProgram(cl_kernel* kernel_ptr,
+                                        const cl_context context,
+                                        const cl_device_id device,
+                                        const unsigned int testId,
+                                        const std::string& testname)
+{
+    auto gen_char = [&]() {
+        static const char dict[] = {
+            " !#$&()*+,-./"
+            "123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`"
+            "abcdefghijklmnopqrstuvwxyz{|}~"
+        };
+        return dict[genrand_int32(gMTdata) % ((int)sizeof(dict) - 1)];
+    };
+
+    std::array<std::vector<std::string>, 2> formats = {
+        { { "%f", "%e", "%g", "%a", "%F", "%E", "%G", "%A" },
+          { "%d", "%i", "%u", "%x" } }
```

Comment:
Do we want `%o` and `%X` too?

Reply:
Done

Reply:
I see you added `%o` but what about `%X` (upper case `X`)?

Reply:
oops, I confused upper/lower case, corrected